software/contrib/cvecorder.py

```diff
@@ -0,0 +1,348 @@
+
+output_1: CV record / playback
+output_2: CV record / playback
+output_3: CV record / playback
+output_4: CV record / playback
+output_5: CV record / playback
+output_6: CV record / playback
+
+'''
+
+'''
+To do:
+- Add a morph capability using k2: left reduces CV values, right increases. Don’t adjust all values at once, so odds, then evens with each slight movement
+- Create documentation
+- Create upstream pull request
+'''
+
+# Needed if using europi_script
+#class CVecorder(EuroPiScript):
+class CVecorder():
+    def __init__(self):
+        
+        # Needed if using europi_script
+        super().__init__()
+
+        # Overclock the Pico for improved performance.
+        machine.freq(250_000_000)
+
+        # Initialize variables
+        self.step = 0
+        self.stepLength = 64
+        self.clockStep = 0
+        self.ActiveCvr = 0
+        self.ActiveBank = 0
+        self.resetTimeout = 500
+        self.debug = True
+        self.CvIn = 0
+        self.bankToSave = 0
+        self.debugTest = False
+
+        self.numCVR = 5  # Number of CV recorder channels - zero based
+        self.numCVRBanks = 5  # Number of CV recording channel banks - zero based
+
+        #self.CVR = []  # CV recorder channels
+        #self.CvRecording = []  # CV recorder flags
+
+        # Load CV Recordings from a previously stored state on disk or initialize if blank
+        self.loadState()
+
+        # Test routine, pick a random bank n times and save, then load the state
+        if self.debugTest:
+            print(micropython.mem_info("level"))
+            for n in range(3000):
+                # Clear vars
+                #self.CvRecording = []
+                print(f"Running test: {n}")
+                self.ActiveBank = randint(0, self.numCVRBanks)
+                self.ActiveCvr = randint(0, self.numCVR)
+                for i in range(0, self.stepLength-1):
+                    self.CVR[self.ActiveBank][self.ActiveCvr][i] = uniform(0.0, 9.99)
+                    #print(f"[{self.ActiveBank}][{self.ActiveCvr}][{i}] = {self.CVR[self.ActiveBank][self.ActiveCvr][i]}")
+                self.bankToSave = self.ActiveBank
+                self.saveState()
+                self.loadState()
+
+        @din.handler
+        def dInput():
+            self.handleClock()
+            self.clockStep +=1
+
+        @din.handler_falling
+        def endClock():
+            self.handleClock()
+
+        @b1.handler
+        def b1Pressed():
+            # Set recording boolean to true and clear the recording buffer
+            self.CvRecording[self.ActiveCvr] = 'pending'
+            # Clear the array
+            for n in range (0, self.stepLength):
+                self.CVR[self.ActiveBank][self.ActiveCvr][n] = 0
+
+        # B2 Long press
+        @b2.handler_falling
+        def b2PressedLong():
+            # 2000ms press clears all banks
```

Comment:
I tried a slightly different approach with CVecorder compared to Consequencer as I wanted the most used function to trigger before the button was released. TBH, it may be slightly easier if there was a long_press(num_seconds) function in the firmware. I feel that this function should also make sure it reverts any change that was made as a result of the short press though too.